pmtiles/bucket.go

```diff
@@ -2,64 +2,117 @@ package pmtiles
 
 import (
 	"context"
+	"errors"
 	"fmt"
-	"gocloud.dev/blob"
 	"io"
 	"net/http"
 	"net/url"
 	"os"
 	"path"
 	"path/filepath"
 	"strings"
+
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"gocloud.dev/blob"
 )
 
 // Bucket is an abstration over a gocloud or plain HTTP bucket.
 type Bucket interface {
 	Close() error
 	NewRangeReader(ctx context.Context, key string, offset int64, length int64) (io.ReadCloser, error)
+	NewRangeReaderEtag(ctx context.Context, key string, offset int64, length int64, etag string) (io.ReadCloser, string, error)
+}
+
+type RefreshRequiredError struct {
```

Comment:
I think it will always be 412, we will never see 416 since 412 should take priority. Unless we want to handle misbehaving servers (like R2 HTTP API right now). Should we consolidate this internally by not bothering to store a status code?

Reply:
The status code is only used for the error message for debugging, we could make it `statusCode` so no one can depend on the value?